### DISCUSSION TITLE
No module named 'MySQLdb' on ARM

### DISCUSSION THREAD

**Discussion by hanrongMan** [apache/airflow#25831] [Q&A] [answered]
I installed by docker-compose for airflow version 2.3.3,it happens this error when I edit connection of mysql and test it. 

I used airflow info command ,then show the picture below. it had apache-airflow-providers-mysql,but I do not know why it happened,please help me.
<img width="444" alt="image" src="https://user-images.githubusercontent.com/19281198/185657002-6d04c010-c82e-41c0-a8ac-2e35e421ae29.png">

**Comment by hanrongMan** [comment]
![image](https://user-images.githubusercontent.com/19281198/185657745-ccb9a07c-ce76-4eb9-bd11-5018c700e282.png)

**Comment by potiuk** [comment, marked as answer]
Likely you are running it on M1 machine. Docker image on M1 only supports Postgres.

**Reply by hanrongMan** [reply]
Thank you very much for answering my question. 
Indeed,I am running  chip  Apple M1 Pro machine in MacBook Pro.
Is there an image that supports MySQL?

**Reply by potiuk** [reply]
no

**Reply by potiuk** [reply]
More detailed answer. No currently there is no supported way. But this might change in the future (but it has rather low priority):

* maybe there will be run if mysql start supporting  Debian installation of MySQL client:
https://www.mysql.com/support/supportedplatforms/database.html

* You can also try to build your own image based on CentOS. We also might want to support Centos image in the future https://github.com/apache/airflow/pull/24825 - but this is not a high priority.

* You can also try to change platform in your docker-compose. I am not sure whether it will work but if you set x86 as platform -  https://docs.docker.com/compose/compose-file/compose-file-v2/#platform, mysql might start working. It will be almost unusable - Emulation on Mac slows down airflow 16x times or more and your CPUs will start spinning so this might be not usable. But you can try (I do not recommend it though and I have not tried it).

**Reply by hanrongMan** [reply]
ok,thank you.

**Reply by eugeneYWang** [reply]
@potiuk Hi Jarek, I ran into this issue and Google guided me to here.

As you referred to mysql's supporting link, I checked its latest status and it seems like MySQL has supported Debian installation as of Jan 16 2023. Is there a plan to support Apple M1 platform? 

> maybe there will be run if mysql start supporting Debian installation of MySQL client:
https://www.mysql.com/support/supportedplatforms/database.html

**Reply by potiuk** [reply]
Where do. you see that they support ARM installation on Debian @eugeneYWang ? 

From what I see in https://www.mysql.com/support/supportedplatforms/database.html it is still on x86_64 only:

<img width="1094" alt="Screenshot 2023-01-17 at 07 53 33" src="https://user-images.githubusercontent.com/595491/212829846-910a3f6f-b1e6-41bb-8b9f-b1c9418a2093.png">

**Reply by eugeneYWang** [reply]
@potiuk I misunderstood it. Thank you for clarifying. 

Are you recently seeing any possibility for Airflow to support MYSQL hook in Apple M1 platform? It could help a lot of people.

**Reply by Taragolis** [reply]
1.  Build by your own all system dependencies which actually required a lot of RAM, Time to complete, patch with specific version of boost.
2. Try to upstream this problem to Oracle

**Reply by Taragolis** [reply]
You also could try to build `mysqlclient` by use MariaDB development headers which included in most of Linux distr like Debian: see [default-libmysqlclient-dev](https://packages.debian.org/bullseye/default-libmysqlclient-dev), but I do not know is it fully compatible with MySQL. 

So you could experiment:

```console

❯ sudo apt-get update
❯ sudo apt-get install default-libmysqlclient-dev build-essential
❯ pip install mysqlclient
```

**Reply by potiuk** [reply]
> Are you recently seeing any possibility for Airflow to support MYSQL hook in Apple M1 platform? It could help a lot of people.

First of all I am not sure if this is for "a lot" people and if it is a problem at all. ARM support for Airlfow is experimental and Apple M1 as platform to run Airlfow is really only useful for local development  - and for local development, there is virtually no difference to use Postgres vs. MySQL for Airflow. The only REAL difference is production use on ARM - which I believe is rather limited number of people. Also we **STRONGLY** recommend people to use Postgres as production backend, and if you are considering ARM as your production platform (likely you do not have legacy ARM Airlfow running) - choosing Postgres is a good idea in general.

If anything - I consider lack of ARM support for MySQL a good incentive for people to switch to Postgres. And taking into account how many problems we have with MySQL vs. Postgres, I have virtually no incentive to go extra mile and spend time on supporting ARM + MySQL if ARM + Postgres is a viable alternative.

And coming back to options - It's really the problem of Oracle not supporting it for Debian. Since our reference image is based on Debian, the only possible approach without Oracle adding Debian support is one of:

1. Trying to build our own mysql client. I definitely do not want it. There might be many issues with that and I do not want Airflow commun be responsible for diagnosing and fixing errors that might come up.

2. Adding Centos as base image option for our reference image - this has been already discussed and we have some draft attempts https://github.com/apache/airflow/pull/24825 . This **might** actualy get implemented some time soon-ish if I have a bit of time for that. It will add some complexity to our CI/Build process, but It might be a viable approach - and side effect **might** be that we will support MySQL. But supoorting MySQL on Apple M1 docker image for development is not - on its own - strong enough incentive to do it. If we get it as a side effect of CentOS image as a base, it might happen.

**Comment by Ibraitas** [comment]
Works fine (mysql hooks) for me with additional *platform* parameter. m1 pro 2021, but postgres backend
![image](https://github.com/apache/airflow/assets/45786968/7124b28e-66b4-4654-ba9b-9b5accc4333c)
```docker
platform: linux/x86_64
```

**Reply by eugeneYWang** [reply]
Yeah, I found it is somehow fixed. No issue connecting to mysql as well

**Reply by Taragolis** [reply]
This one resolved for a long time ago, somewhere between 2.5.1 and 2.6.0 in docker image was added support of MySQL backend for ARM images. That mean it also build and pre-install [`mysqlclient`](https://pypi.org/project/mysqlclient/) which also uses in hooks by default

**Reply by Taragolis** [reply]
I've checked, in 2.6.0

**Reply by Taragolis** [reply]
https://github.com/apache/airflow/pull/38267
